Daily rewards for active nodes (#1064)
hashgraph/hedera-improvement-proposal: HIP/hip-1064.md

hip: 1064
title: Daily Rewards For Active Nodes
author: Neeharika Sompalli <@Neeharika-Sompalli>
working-group: Leemon Baird (@lbaird), Richard Bair <@rbair23>, Michael Tinker <@tinker-michaelj>
requested-by: Hashgraph
type: Standards Track
category: Core
needs-council-approval: Yes
status: Draft
last-call-date-time: 2024-04-08T07:00:00Z
created: 2024-10-17
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/1063
updated: 2024-10-17
---

## Abstract

This HIP proposes a reward mechanism that will incentivize nodes to remain active on the network. The definition of active node is described below. The minimum reward per day for nodes will be a configurable property decided by the governance of the deployed network. This minimum reward may be adjusted based on fees already paid throughout the day.

## Motivation

The goal is to give a strong incentive for node operators to keep them active and properly configured throughout every staking period.

## Rationale

Clearly defining the criteria for node activity, we can ensure that rewards are distributed fairly, incentivizing nodes to remain active and contribute to the network’s stability and growth.

> ### Terminology

Author: I looked at other HIPS and saw this is different in each HIP. I found HIP-904 had it in Rationale. Changed it

In the hashgraph algorithm, honest nodes that are online and contributing to consensus will regularly create unique famous witnesses, called judges, in nearly every round.

This HIP defines a node as *active* if it creates a judge in a significant fraction of rounds during a staking period, for example 80%. Any node that is trustworthy and submits events will create judges. If the node doesn’t submit events or when other nodes don’t build on this node’s events ,  the node will not create judges.

> **User Stories**

Author: yes missed that. Thanks Joseph

The implementation will be tested under various scenarios, including:

> - As a node operator, I will expect to get a minimum node reward when the node is active for at least `nodes.activeRoundsPercent`  rounds
> - As a node operator, I will not expect to get any node rewards when the node is not active for at least `nodes.activeRoundsPercent`  rounds
> - As a network operator, when `nodes.adjustNodeFees` is set to true, each nodes gets reward reduced by the average node fees already collected during that period.
> - As a network operator, when `nodes.adjustNodeFees` is set to false, node rewards are not reduced by the average node fees already collected during that period.
> - As  a network operator, each active node will get minimum reward defined by `nodes.minNodeReward`

Reviewer: ```suggestion
- As a node operator, I expect to receive at least a minimum node reward when the
  node is active for at least `nodes.activeRoundsPercent` rounds
- As a node operator, I do not expect to receive any node rewards when the node
  is not active for at least `nodes.activeRoundsPercent` rounds
- As a network operator, when `nodes.adjustNodeFees` is set to true, I expect
  each node reward to be reduced by the average node fees already collected
  during that period.
- As a network operator, when `nodes.adjustNodeFees` is set to false, I do not
  expect each node reward to be reduced by the average node fees already
  collected during that period.
- As a network operator, I expect each active node to receive at least a
  minimum reward defined by `nodes.minNodeReward`
```

## Specification

> There will be three configurable properties added, which can be changed using `FileUpdate` transaction signed by the governance of the deployed network.

Reviewer: ```suggestion
There will be three configurable properties added, which can be changed using a `FileUpdate` transaction signed by the governance of the deployed network.
```

Author: Updated to remove how the properties are updated as Michael H mentioned below.

> There will be three configurable properties added, which can be changed using `FileUpdate` transaction signed by the governance of the deployed network.

Reviewer 2: I think we should not mention how a property is changed. This may change in the future.

> - `nodes.minNodeReward` : The minimum daily node reward amount per day
> - `nodes.adjustNodeFees` : A flag if set, will reduce the rewards given for a staking period by the average node fees already collected during that period.
> - `nodes.activeRoundsPercent` : The percent number of rounds a node must create a judge to be considered active. This is set to 80% by default.

Reviewer: ```suggestion
- `nodes.minNodeReward` : A minimum daily node reward amount. Each node SHALL
  receive at least this many tinybar each day that the node is "Active".
- `nodes.adjustNodeFees` : A flag to adjust node fees. If this is set, the
  network SHALL reduce the rewards given for a staking period by the average
  node fees already collected during that period.
- `nodes.activeRoundsPercent` : A percentage value relating to active nodes.
  A node MUST create a judge in at least this percentage of rounds to be
  considered active. This is set to 80% by default.
```

Note that the reward is adjusted by same amount for all nodes when `staking.fees.adjustNodeFees` is true. I.e., Alice’s node is paid the constant reward minus the average node fees received that day, averaged across all nodes. We will NOT subtract the node fees earned by Alice in particular.

The node rewards are distributed at the end of a staking period using CryptoTransfer transactions from `ledger.fundingAccount` to each of the node accounts. Since each CryptoTransfer only allows `ledger.transfers.maxLen` number of transfers, based on the it is possible there will be multiple CryptoTransfers that will distribute rewards based on number of node accounts in the network.

This HIP modifies the following state protobufs.

> - Currently PlatformState’s ConsensusSnapshot contains information about the `judge_hashes`. But, it doesn’t provide the creator ids. We will deprecate the existing `judge_hashes`  and add a list of `judge_infos` . Each `JudgeInfo` will include the creator node id and the judge hash for that node. Any node that didn’t create a judge in a round will not be present in the list.

Reviewer: ```suggestion
- Currently PlatformState’s ConsensusSnapshot contains information about the `judge_hashes`. But, it doesn’t provide the creator ids. We will deprecate the existing `judge_hashes`  and add a list of `judges`. Each `Judge` will include the creator node id and the judge hash for that node. Any node that didn’t create a judge in a round will not be present in the list.
```

```
/**
 * A consensus snapshot.<br/>
 * This is a snapshot of the consensus state for a particular round.
 *
 * This message SHALL record consensus data necessary for restart
 * and reconnect.
 */
message ConsensusSnapshot {
    ....
    /**
     * A list of SHA-384 hash values.<br/>
     * The hashes of all judges for this round.
     * <p>
     * This list SHALL be ordered by creator ID.<br/>
     * This list MUST be deterministically ordered.
     */
    repeated bytes judge_hashes = 2 [deprecated=true];
    ....
    /*
     * A list of judges.
     * This list SHALL be ordered by creator ID.<br/>
     * This list MUST be deterministically ordered.
     */
    repeated Judge judges = 6;
}

/**
 * A judge information
 */

> message Judge {
>   /**
>    * The creator node ID who created this judge.
>    */
>     uint64 creator_id = 1;
>     /**
>      * SHA-384 hash value of this judge
>      */
>     bytes judge_hash = 2;
> }

Reviewer: Indentation and spacing were a little wonky here
```suggestion
message Judge {
    /**
     * The creator node ID who created this judge.
     */
    uint64 creator_id = 1;

    /**
     * SHA-384 hash value of this judge
     */
    bytes judge_hash = 2;
}
```

```

> - To store the number of rounds in a staking period and the number of rounds in which each node created judges, a new singleton state `NodeRewards`will be added to `TokenService` . This singleton state will be updated at the end of every round.
> 

Reviewer: Spacing
```suggestion
- To store the number of rounds in a staking period and the number of rounds in which each node created judges, a new singleton state `NodeRewards` will be added to `TokenService`. This singleton state will be updated at the end of every round.

```

```
syntax = "proto3";

> package proto;

Reviewer: package for new protos should not be `proto`.
```suggestion
package com.hedera.hapi.node.state.token;
```

option java_package = "com.hederahashgraph.api.proto.java";
// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package
option java_multiple_files = true;

> /**
>  * A node rewards message. This is used to find number of active nodes in a staking period based on number of judges
>  * each node created in that period.
>  */
> message NodeRewards {
>     /**
>      * Number of rounds in a staking period.
>      */
>     uint64 num_rounds_in_staking_period = 1;
>     /**
>      * Number of judges created by each node in a staking period.
>      */
>     repeated NodeActivity node_activities = 2;
> }
> /**
>   * A node activity containing number of judges created in a staking period.
>  */
> message NodeActivity {
>     /**
>      * Node ID of the node.
>      */
>     uint64 node_id = 1;
>     /**
>      * Number of judges created by the node.
>      */
>     uint64 num_judge_rounds = 2;
> }

Reviewer: Specification updates
```suggestion
/**
 * A record of node "active" status.<br/>
 * This is used to record the number of "active" nodes in a staking
 * period based on number of judges each node created in that period.
 *
 * A Node SHALL be considered "active" if it produced "judges" according
 * to the consensus algorithm in a percentage of rounds, during the
 * staking period, greater than the network configuration value for
 * `nodes.activeRoundsPercent`.
 */
message NodeRewards {
    /**
     * A number of rounds in the staking period.
     */
    uint64 num_rounds_in_staking_period = 1;

    /**
     * A list of node activities.<br/>
     * This records the number of rounds when each node created
     * judges for the consensus algorithm.
     * <p>
     * This list SHALL contain one entry for each node participating
     * in consensus during the staking period.
     */
    repeated NodeActivity node_activities = 2;
}
/**
 * A record of judge rounds for a single node.<br/>
 * This records, for a single node, the number of rounds during a staking
 * period that contained at least one judge created by that node.
 *
 * This message SHALL NOT record the total number of rounds in a staking
 * period.<br/>
 * This message SHALL record a count of rounds for a single node.<br/>
 * This message SHALL contain the number of rounds that contain at
 * least one judge from the node identified.
 */
message NodeActivity {
    /**
     * A node identifier.
     */
    uint64 node_id = 1;

    /**
     * A count of rounds.<br/>
     * This is the count of rounds in which the node identified
     * by `node_id` created judges.
     */
    uint64 num_judge_rounds = 2;
}
```

```

## Backward Compatibility

Existing nodes will not require any changes to their software or configuration to be eligible for rewards, as long as they meet the active criteria.

## Security Implications

No known security concerns.

## How to Teach This

This HIP incentivizes node operators for keeping the node active by guaranteeing a minimum reward in the periods that they accomplish this.

## Rejected Ideas

> - Considered using number of submitted transactions to node to treat the node as active.

Reviewer: ```suggestion
- We considered using the number of transactions submitted by a node to assess node activity.
   - This was rejected in favor of counting rounds with judges because the latter
     approach incorporates consensus network assessment of node "honesty" and
     is more "fair".
```

---
hip: 1064
title: Daily Rewards For Active Nodes
author: Neeharika Sompalli <@Neeharika-Sompalli>
working-group: Leemon Baird (@lbaird), Richard Bair <@rbair23>, Michael Tinker <@tinker-michaelj>
requested-by: Hashgraph
type: Standards Track
category: Core
needs-council-approval: Yes
status: Draft
last-call-date-time: 2024-04-08T07:00:00Z
created: 2024-10-17
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/1063
updated: 2024-10-17
---

## Abstract

This HIP proposes a reward mechanism that will incentivize nodes to remain active on the network. The definition of active node is described below. The minimum reward per day for nodes will be a configurable property decided by the council. This minimum reward may be adjusted based on fees already paid throughout the day.

> 

Reviewer 3: In the context of Hiero, we should reword "by the council" to mean "by the governance of the deployed network", or something similar.

Author: Changed. Thanks

## Motivation

The goal is to give a strong incentive for node operators to keep them active and properly configured throughout every staking period.

## Rationale

Clearly defining the criteria for node activity, we can ensure that rewards are distributed fairly, incentivizing nodes to remain active and contribute to the network’s stability and growth.

### Terminology

In the Hashgraph consensus algorithm, nodes communicate by sharing information through events. An event contains some transactions.

> In each round, each node creates a *witness* which is an event created in that round that does not have any descendants from the same node in that round. A *judge* is a specific type of witness event that gains importance because it is known and agreed upon by a majority of the network.

Reviewer 3: I did not understand this statement:

> In each round, each node creates a *witness* which is an event created in that round that does not have any descendants from the same node in that round

Maybe it is too much detail in this HIP to define exactly what a witness or a judge is. Perhaps we can just say something like:

"In the hashgraph algorithm, honest nodes that are online and contributing to consensus will regularly create unique famous witnesses, called judges, in nearly every round." That might be sufficient and accurate enough (although we should definitely review the language with Leemon).

Author: Updated as per the suggestion. Thanks Richard

Author: Confirmed with Leemon and updated the definition as per glossary.

This HIP defines a node as *active* if it creates a judge in a significant fraction of rounds during a staking period, for example 80%. Any node that is trustworthy and submits events will create judges. If the node doesn’t submit events or when other nodes don’t build on this node’s events ,  the node will not create judges.

**User Stories**

The implementation will be tested under various scenarios, including:

> - As a node operator, I will expect to get a minimum node reward when the node is active for atleast `nodes.activeRoundsPercent`  rounds

Reviewer 3: ```suggestion
- As a node operator, I will expect to get a minimum node reward when the node is active for at least `nodes.activeRoundsPercent`  rounds
```

Some other lines also should have "at least" as two words instead of 1

Author: Yes changed them . thanks Richard

  collected during that period.
- `nodes.activeRoundsPercent` : A percentage value relating to active nodes.
  A node MUST create a judge in at least this percentage of rounds to be considered active.
  This is set to 80% by default.

The reward SHALL be adjusted based on average fees paid for all nodes when `staking.fees.adjustNodeFees` is true.
Each node is paid the constant reward minus the average node fees received that day.
This reduction is averaged across all nodes; the network SHALL NOT subtract the node fees earned by one node in
particular.

The node rewards SHALL be distributed at the end of a staking period using`CryptoTransfer` transactions from
`ledger.fundingAccount` to each node account. Each `CryptoTransfer` only allows `ledger.transfers.maxLen`
transfers per transaction; it is possible, therefore, that multiple `CryptoTransfer` transaction will be necessary to
distribute rewards, based on the total number of node accounts in the network.

This HIP modifies the following state protobufs.

- Currently, PlatformState’s ConsensusSnapshot contains information about the `judge_hashes`. But, it doesn’t provide

>   the

Author: Fixed. Thanks!

---
hip: 1064
title: Daily Rewards For Active Nodes
author: Neeharika Sompalli <@Neeharika-Sompalli>
working-group: Leemon Baird (@lbaird), Richard Bair <@rbair23>, Michael Tinker <@tinker-michaelj>
requested-by: Hashgraph
type: Standards Track
category: Core
needs-tsc-approval: Yes
status: Draft
last-call-date-time: 2024-04-08T07:00:00Z
created: 2024-10-17
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/1063
updated: 2024-10-17
---

## Abstract

This HIP proposes a reward mechanism that will incentivize nodes to remain active on the network. The definition of
active node is described below. The minimum reward per day for nodes will be a configurable property decided by the
governance of the deployed network. This minimum reward may be adjusted based on fees already paid throughout the day.

### Terminology

<dl>
<dt>Judge</dt>

> <dd>An event that wins the election to be made a judge. It must be a witness, and it will have tended to have been 
> gossiped to most of the other nodes quickly (otherwise it would have lost the election). An event reaches consensus
> when it is an ancestor of all judges in a given round. The first round where that happens is its consensus round. 
> It’s a math theorem that every round is guaranteed to have at least one judge, and a math conjecture that every round
> is guaranteed to have judges created by a supermajority of nodes (>2/3 of weight).</dd>

Reviewer: This confuses me a bit.
Does the following express the same concepts?
```suggestion
<dd>An event wins an election to be made a judge. To win the election an event
must be a witness, and it will tend to be gossiped to most of the other nodes
quickly (otherwise it would have lost the election).

An event reaches consensus when it is an ancestor of all judges in a given round.
The first round where that happens is its consensus round. It is a mathematic
theorem that every round is guaranteed to have at least one judge, and a
mathematic conjecture that every round is guaranteed to have judges created
by a supermajority of nodes (>2/3 of consensus weight).</dd>
```

One thing that is missing:
What is the actual _criteria_ for "winning" the election?
Is it the first set of events that qualify as witnesses for the round?
Is it the first set of events that qualify as "famous witnesses"?

Reviewer: Definitely confirm.  It's possible the glossary has some awkward wording, and I could easily be incorrect here.  The judges concepts are not intuitively clear, so it's best to confirm wording with our hashgraph experts before committing to it.

<dt>Active Node</dt>
<dd>A node is an *Active Node* if it creates a "judge" in a significant fraction of rounds during a staking period, 
for example 80%. Any node that is trustworthy and submits events will create judges. If the node doesn’t submit events
or when other nodes don’t build on this node’s events, the node will not create judges.</dd>
</dl>

## Motivation

The goal is to give a strong incentive for node operators to keep them active and properly configured throughout every
staking period.

## Rationale

Clearly defining the criteria for node activity, we can ensure that rewards are distributed fairly, incentivizing nodes
to remain active and contribute to the network’s stability and growth.

## User Stories

The implementation will be tested under various scenarios, including:

- As a node operator, I expect to receive at least a minimum node reward when the node is active for at least
  `nodes.activeRoundsPercent` rounds

> - As a node operator, I do not expect to receive any node rewards when the node is not active for at least

Reviewer 2: I understand this is following the usual pattern for use cases, but I find it weird nevertheless. How about something like:

"As a user of the network, I expect that non-active nodes do not receive any node rewards..."

Author: > As a user of the network

Do you mean the node operator Michael?

Reviewer 2: No, I meant any regular user. It sounds weird for me to have a use case saying that somebody does not expect a certain behavior. However, as a regular user, I expect that the money I spend is only used for active nodes.

It's not important, though. It is clear what the expected behavior is.

Author: Ah makes sense. Thanks @netopyr . Will add another use case

     * A list of SHA-384 hash values.<br/>
     * The hashes of all judges for this round.
     * <p>
     * This list SHALL be ordered by creator ID.<br/>
     * This list MUST be deterministically ordered.
     */
    repeated bytes judge_hashes = 2 [deprecated=true];
    ....
    /*
     * A list of judges.
     * This list SHALL be ordered by creator ID.<br/>
     * This list MUST be deterministically ordered.
     */
    repeated Judge judges = 6;
}

/**
 * A judge information
 */
message Judge {
    /**
     * The creator node ID who created this judge.
     */
    uint64 creator_id = 1;

    /**
     * SHA-384 hash value of this judge
     */
    bytes judge_hash = 2;
}
```

- To store the number of rounds in a staking period and the number of rounds in which each node created judges, a new
  singleton state `NodeRewards` will be added to `TokenService`. This singleton state will be updated at the end of
  every round.

```
syntax = "proto3";

package com.hedera.hapi.node.state.token;

option java_package = "com.hederahashgraph.api.proto.java";
// <<<pbj.java_package = "com.hedera.hapi.node.state.token">>> This comment is special code for setting PBJ Compiler java package

> option java_multiple_files = true;

Reviewer 2: We should omit the header.

Author: Fixed. Thanks!
